### PR TITLE
Pack invocation field

### PR DESCRIPTION
**Pull request by olorin**
! @markhibberd
/jury approve @markhibberd @ambiata-ci

### CODE REVIEW

File: src/Tinfoil/AEAD/AESGCM/Data.hs [hunk 1]
```diff
@@ -43,6 +45,11 @@ data InvocationField =
     InvocationField !RandomField !InvocationCount
   deriving (Eq, Show)
 
+packInvocationField :: InvocationField -> Word64
+packInvocationField (InvocationField (RandomField rf) (InvocationCount ic)) =
+  -- ic is guaranteed to be non-negative and less than 2^32.
+  ((fromIntegral rf :: Word64) `shiftL` 32) .|. (fromIntegral ic :: Word64)
+
```
Comment (erikd-ambiata) [inline src/Tinfoil/AEAD/AESGCM/Data.hs]: I notice that:
```
newtype InvocationCount =
  InvocationCount {
    unInvocationCount :: Integer
  } deriving (Eq, Show, Ord, Num)
```
Is there any reason that isn't a `Word64` instead of `Integer`. That would avoid the need for `fromIntegral`.

File: test/Test/Tinfoil/AEAD/AESGCM/Data.hs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import           Test.Tinfoil.Arbitrary ()
+
+import           Tinfoil.AEAD.AESGCM.Data
+
+unpackInvocationField :: Word64 -> InvocationField
+unpackInvocationField w =
+  let
+    rf = RandomField . fromIntegral $ w `shiftR` 32
+    ic = InvocationCount . fromIntegral $ w .&. 0x00000000ffffffff
+  in
```
Comment (erikd-ambiata) [inline test/Test/Tinfoil/AEAD/AESGCM/Data.hs]: Relatedly you have an `Integer` there but only seem to be using the bottom 32 bits.

Comment (olorin) [inline test/Test/Tinfoil/AEAD/AESGCM/Data.hs]: Yeah, this is just me being lazy (in a paranoid way) about the possibility of overflow. You're right, I will rework this to be more efficient.

Comment (olorin) [inline test/Test/Tinfoil/AEAD/AESGCM/Data.hs]: @erikd-ambiata how's the current version?

Comment (erikd-ambiata) [inline test/Test/Tinfoil/AEAD/AESGCM/Data.hs]: Much better :).

File: src/Tinfoil/AEAD/AESGCM/Data.hs [hunk 3]
```diff
@@ -34,7 +36,7 @@ newtype RandomField =
 -- startup.
 newtype InvocationCount =
   InvocationCount {
-    unInvocationCount :: Integer
+    unInvocationCount :: Word64
```
Comment (markhibberd) [inline src/Tinfoil/AEAD/AESGCM/Data.hs]: Isn't this only 32 bits?

Comment (markhibberd) [inline src/Tinfoil/AEAD/AESGCM/Data.hs]: Sorry, I see this is intentional from what you spec'd, it isn't quite what I expected, but I think I understand why. I think I would of checked and made sure it fit in 32 bits and left the type as Word32 here to make sure.

Comment (olorin) [inline src/Tinfoil/AEAD/AESGCM/Data.hs]: Yep that was the other option I considered, used a `Word64` at @erikd-ambiata's suggestion to avoid the `fromIntegral`. I do agree that a `Word32` would be clearer though, might go with that.

Comment (olorin) [inline src/Tinfoil/AEAD/AESGCM/Data.hs]: Updated, and added an extra test for overflow.

Comment (erikd-ambiata) [inline src/Tinfoil/AEAD/AESGCM/Data.hs]: :+1:

**Comment by markhibberd** [comment]
👍
